src/commands/watch.rs

```diff
@@ -87,54 +88,81 @@ impl WatchCommand {
             async move {
+                // Map all the events of this action to an effect
+                let mut effects = Effects::new();
                 for event in action.events.iter() {
-                    // Exit if interrupt signal sent
                     if event.signals().any(|s| s.eq(&Interrupt)) {
-                        action.outcome(Outcome::both(Outcome::Stop, Outcome::Exit));
-                        return Ok::<(), Infallible>(());
+                        effects.add(Effect::Exit);
                     }
-
-                    if event.paths().any(|(p, _)| p.ends_with(app_manifest_path.clone())) {
+                    if event
+                        .tags
+                        .iter()
+                        .any(|t| matches!(t, Tag::ProcessCompletion(Some(ProcessEnd::Success))))
+                    {
+                        effects.add(Effect::ProcessCompleted);
+                    }
+                    if event.tags.iter().any(|t| {
+                        matches!(
+                            t,
+                            Tag::ProcessCompletion(None)
+                                | Tag::ProcessCompletion(Some(
+                                    ProcessEnd::Exception(_)
+                                        | ProcessEnd::ExitError(_)
+                                        | ProcessEnd::ExitSignal(_)
+                                        | ProcessEnd::ExitStop(_)
+                                ))
+                        )
+                    }) {
+                        effects.add(Effect::ProcessFailed);
+                    }
+                    if filter_handle.matches_manifest_pattern(event) {
                         // TODO: Reconfigure the watcher
                         eprintln!("Application manifest has changed. If this included changes to the watch configuration, please restart Spin.");
+                        effects.add(Effect::ManifestChange);
+                    }
+                    if filter_handle.matches_source_pattern(event) {
+                        effects.add(Effect::SourceChange);
+                    }
+                    if filter_handle.matches_artifact_pattern(event) {
+                        effects.add(Effect::ArtifactChange);
                     }
                 }
-
-                action.outcome(Outcome::if_running(
-                    Outcome::both(
-                        Outcome::both(
-                            Outcome::Stop,
-                            match self.clear {
-                                true => Outcome::Clear,
-                                false => Outcome::DoNothing,
-                            },
-                        ),
-                        Outcome::Start,
-                    ),
-                    Outcome::Start,
-                ));
-
-            Ok::<(), Infallible>(())
-        }});
+                action.outcome(state_machine.handle(effects.reduce()));
```

Comment:
I like the use of `Effect` ordering to choose the highest priority effect, but I wonder if you could replace `Effects` with a `Vec<Effect>` and do something like this here:

```suggestion
                // Pick the highest priority effect to apply
                let effect = effects.into_iter().min();
                action.outcome(state_machine.handle(effect));
```

The `Effects` abstraction is nice, but its only used in this one place so maybe not worth the extra code and indirection.

Reply:
Think I'm going to keep this one. I like making the dependency on the ordering very explicit and built into `Effects`. It also means that we can make sure that this ordering isn't subtly broken by unit testing `Effects`.